File: lib/domain/models/published_schedule.dart
import 'deck.dart';
import 'scheduled_vocab.dart';

/// One entry of a published queue: what to draw, and when.
///
/// Deliberately not a [ScheduledVocab]. That carries a whole [Vocab], most of
/// which the lock screen has no use for, and the payload has to survive a
/// round trip through JSON — which a full term, with its deck and its
/// timestamps, would not do without inventing fields nobody reads.
class PublishedEntry {
  const PublishedEntry({
    required this.showAt,
    required this.vocabId,
    required this.term,
    required this.translation,
  });

  /// Reduces a scheduled term to what the lock screen actually draws.
  factory PublishedEntry.of(ScheduledVocab scheduled) => PublishedEntry(
    showAt: scheduled.showAt,
    vocabId: scheduled.vocab.id,
    term: scheduled.vocab.term,
    translation: scheduled.vocab.translation,
  );

  final DateTime showAt;

  /// Which term this was. Kept so the app can work out, on its next run,
  /// which entries have already had their turn.
  final String vocabId;

  final String term;
  final String translation;

  Map<String, Object?> toJson() => {
    'showAt': showAt.millisecondsSinceEpoch,
    'vocabId': vocabId,
    'term': term,
    'translation': translation,
  };

  factory PublishedEntry.fromJson(Map<String, Object?> json) => PublishedEntry(
    showAt: DateTime.fromMillisecondsSinceEpoch(
      json['showAt']! as int,
      isUtc: true,
    ),
    vocabId: json['vocabId']! as String,
    term: json['term']! as String,
    translation: json['translation']! as String,
  );

  @override
  bool operator ==(Object other) =>
      other is PublishedEntry &&
      other.showAt == showAt &&
      other.vocabId == vocabId &&
      other.term == term &&
      other.translation == translation;

  @override
  int get hashCode => Object.hash(showAt, vocabId, term, translation);

  @override
  String toString() => 'PublishedEntry($showAt, $vocabId)';
}

/// The queue of upcoming terms as the native side receives it.
///
/// This is the contract between the Flutter app and the lock screen: the app
/// writes one of these whenever it knows something new, and the widget reads
/// it without any Dart running. Everything the widget has to draw is therefore
/// already in here — it cannot call back to ask.
///
/// The app reads it back too, to work out which entries have had their turn
/// since it last ran.
///
/// Instants are epoch milliseconds rather than formatted dates, because both
/// platforms read that without agreeing on a date format first.
class PublishedSchedule {
  const PublishedSchedule({
    required this.generatedAt,
    required this.deckId,
    required this.deckName,
    required this.sourceLanguage,
    required this.targetLanguage,
    required this.interval,
    required this.entries,
  });

  /// Builds a payload from a freshly computed schedule.
  factory PublishedSchedule.of({
    required DateTime generatedAt,
    required Deck deck,
    required List<ScheduledVocab> upcoming,
  }) => PublishedSchedule(
    generatedAt: generatedAt,
    deckId: deck.id,
    deckName: deck.name,
    sourceLanguage: deck.sourceLanguage,
    targetLanguage: deck.targetLanguage,
    interval: deck.displayInterval,
    entries: upcoming.map(PublishedEntry.of).toList(),
  );

  /// The shape of [toJson].
  ///
  /// Written into every payload so a reader built against an older shape can
  /// refuse one it does not understand instead of misreading it. Bump this
  /// whenever a field changes meaning or disappears.
  static const version = 1;

  /// When the app last worked the queue out. A widget can use this to tell a
  /// stale queue from a current one.
  final DateTime generatedAt;

  final String deckId;
  final String deckName;
  final String sourceLanguage;
  final String targetLanguage;

  /// The gap between two consecutive entries, repeated here so the widget does
  /// not have to derive it from the timestamps.
  final Duration interval;

  /// Ordered by [PublishedEntry.showAt], earliest first.
  final List<PublishedEntry> entries;

  /// The entry the lock screen is showing at [now], or null when the queue
  /// does not cover that moment.
  ///
  /// Null means there is nothing to carry on from: either the queue lies
  /// wholly in the future, or it ran out while the app was away.
  PublishedEntry? currentAt(DateTime now) {
    for (final entry in entries.reversed) {
      if (entry.showAt.isAfter(now)) continue;

      // The last entry that has started. Whether it is still running decides
      // whether the queue reaches [now] at all.
      return now.isBefore(entry.showAt.add(interval)) ? entry : null;
    }

    return null;
  }

  /// The entries whose slot is completely over by [now].
  ///
  /// The entry still showing is deliberately not among them. It is having its
  /// turn, not finished with it — counting it now would cut that turn short,
  /// because a term counted as shown falls behind in the next queue.
  List<PublishedEntry> finishedBy(DateTime now) => entries
      .where((entry) => !entry.showAt.add(interval).isAfter(now))
      .toList();

  Map<String, Object?> toJson() => {
    'schemaVersion': version,
    'generatedAt': generatedAt.millisecondsSinceEpoch,
    'deckId': deckId,
    'deckName': deckName,
    'sourceLanguage': sourceLanguage,
    'targetLanguage': targetLanguage,
    'intervalMinutes': interval.inMinutes,
    'entries': [for (final entry in entries) entry.toJson()],
  };

  /// Reads a payload back, or throws if it is not one.
  ///
  /// A payload from a newer app than this one is refused rather than guessed
  /// at — the whole point of [version].
  factory PublishedSchedule.fromJson(Map<String, Object?> json) {
    final schemaVersion = json['schemaVersion'];
    if (schemaVersion != version) {
      throw FormatException('Unsupported schedule version: $schemaVersion');
    }

    return PublishedSchedule(
      generatedAt: DateTime.fromMillisecondsSinceEpoch(
        json['generatedAt']! as int,
        isUtc: true,
      ),
      deckId: json['deckId']! as String,
      deckName: json['deckName']! as String,
      sourceLanguage: json['sourceLanguage']! as String,
      targetLanguage: json['targetLanguage']! as String,
      interval: Duration(minutes: json['intervalMinutes']! as int),
      entries: [
        for (final entry in json['entries']! as List<Object?>)
          PublishedEntry.fromJson(entry! as Map<String, Object?>),
      ],
    );
  }
}

File: android/app/src/main/kotlin/com/github/quendo05/nagara/widget/PublishedSchedule.kt
package com.github.quendo05.nagara.widget

import org.json.JSONObject

/**
 * One entry of the published queue: what to draw, and when.
 *
 * Mirrors `PublishedEntry` in `lib/domain/models/published_schedule.dart`.
 * There is no compiler between the two, so a field changed on one side is a
 * field that has to change on this one — [PublishedSchedule.VERSION] is what
 * catches it when that is forgotten.
 */
data class PublishedEntry(
    val showAtMillis: Long,
    val vocabId: String,
    val term: String,
    val translation: String,
)

/**
 * The queue of upcoming terms, as the Flutter app leaves it behind.
 *
 * Read without any Dart running: everything the widget has to draw is already
 * in here, because there is nothing to call back to.
 */
data class PublishedSchedule(
    val generatedAtMillis: Long,
    val deckId: String,
    val deckName: String,
    val sourceLanguage: String,
    val targetLanguage: String,
    val intervalMillis: Long,
    /** Ordered by [PublishedEntry.showAtMillis], earliest first. */
    val entries: List<PublishedEntry>,
) {
    /**
     * The entry that should be on screen at [nowMillis], or null when the
     * queue does not cover that moment — it lies wholly in the future, or it
     * ran out while the app was away.
     *
     * Android has no timeline to hand the system the way WidgetKit does, so
     * the widget answers this itself every time it redraws. That is also what
     * makes it drift-tolerant: an update that arrives four minutes late still
     * renders the term whose slot is running, not the one after it.
     */
    fun currentAt(nowMillis: Long): PublishedEntry? {
        for (entry in entries.asReversed()) {
            if (entry.showAtMillis > nowMillis) continue

            // The last entry that has started. Whether its slot is still
            // running decides whether the queue reaches [nowMillis] at all.
            return if (nowMillis < entry.showAtMillis + intervalMillis) entry else null
        }

        return null
    }

    companion object {
        /**
         * The payload shape this build understands.
         *
         * Must match `PublishedSchedule.version` on the Dart side. A payload
         * written by a newer app is refused rather than guessed at, which is
         * the whole point of carrying it.
         */
        const val VERSION = 1

        /**
         * Reads a payload, or returns null if it is not one this build can use.
         *
         * Null rather than a throw, because there is nobody here to tell. A
         * widget renders its empty state and waits for the next write — the
         * same thing it does before the app has ever published anything.
         */
        fun fromJson(raw: String): PublishedSchedule? = try {
            val json = JSONObject(raw)
            if (json.getInt("schemaVersion") != VERSION) null else read(json)
        } catch (error: Exception) {
            null
        }

        private fun read(json: JSONObject): PublishedSchedule {
            val array = json.getJSONArray("entries")
            val entries = ArrayList<PublishedEntry>(array.length())
            for (index in 0 until array.length()) {
                val entry = array.getJSONObject(index)
                entries.add(
                    PublishedEntry(
                        showAtMillis = entry.getLong("showAt"),
                        vocabId = entry.getString("vocabId"),
                        term = entry.getString("term"),
                        translation = entry.getString("translation"),
                    )
                )
            }

            return PublishedSchedule(
                generatedAtMillis = json.getLong("generatedAt"),
                deckId = json.getString("deckId"),
                deckName = json.getString("deckName"),
                sourceLanguage = json.getString("sourceLanguage"),
                targetLanguage = json.getString("targetLanguage"),
                // Minutes on the wire, because that is the unit a deck's
                // display interval is stored in; milliseconds here, because
                // that is what every Android clock deals in.
                intervalMillis = json.getLong("intervalMinutes") * 60_000L,
                entries = entries,
            )
        }
    }
}
